catkin_home/src/devices/src/InputAudio.cpp

```diff
@@ -1,26 +1,326 @@
-#define BUFFER_SIZE 256
+extern "C" {
+#include "rnnoise.h"
+}
+
+#include "RnnoiseMethods/HelperMethods.hpp"
+
+
 using namespace std;
+
+
+#define BUFFER_SIZE 256
+
+constexpr int NUMBER_FRAMES_RNNOISE = 480;
+constexpr long SAMPLE_RATE = 48000L;
+constexpr int MS_IN_A_CHUNK = (NUMBER_FRAMES_RNNOISE * 1000L) / SAMPLE_RATE;
+
+constexpr int NUMBER_CHUNKS_PAST_RECORDS = 13L;
+constexpr long PAST_RECORD_BUFFER_SIZE = NUMBER_CHUNKS_PAST_RECORDS * NUMBER_FRAMES_RNNOISE;
+
+constexpr int SECONDS_ACTUAL_RECORDING = 10;
+constexpr long ACTUAL_RECORDING_BUFFER_SIZE = SECONDS_ACTUAL_RECORDING * SAMPLE_RATE;
+
+constexpr long RECORDING_BUFFER_SIZE = ACTUAL_RECORDING_BUFFER_SIZE + PAST_RECORD_BUFFER_SIZE;
+
+// Constants first part of algorithm.
+constexpr int MAX_INIT_MEM = 4;
+constexpr int NUM_TO_INIT_VOICE = 4;
+constexpr float MIN_PROB_IN_INIT = 0.85;
+
+// Constants second part of algorithm.
+constexpr int NUM_ITERATIONS_ALMOST_IN_VOICE = 300L / MS_IN_A_CHUNK;
+constexpr float MIN_PROB_IN_ALMOST = 0.85;
+
+// Third part of algorithm.
+constexpr int MAX_END_VOICE_MEM = 5;
+constexpr int MAX_ITERATIONS_WITHOUT_VOICE = 900L / MS_IN_A_CHUNK;
+constexpr int NUM_TO_END_VOICE = 4;
+constexpr float MIN_PROB_IN_END = 0.80;
+
+constexpr int HISTORY_SIZE = MAX_INIT_MEM > MAX_END_VOICE_MEM ? MAX_INIT_MEM : MAX_END_VOICE_MEM;
+
+
 ros::Publisher publi;
 
+DenoiseState *st;
+
+
+void RNNoiseWarmUp(DenoiseState *st) {
+  int i;
+  float x[NUMBER_FRAMES_RNNOISE];
+  
+  FILE *f1;
+  f1 = fopen("./warmup_street_20dB.raw", "r");
+
+  while (true) {
+    int16_t tmp[NUMBER_FRAMES_RNNOISE];
+    fread(tmp, sizeof(int16_t), NUMBER_FRAMES_RNNOISE, f1);
+    if (feof(f1)) break;
+
+    convertFromShortArrayToFloatArray(tmp, NUMBER_FRAMES_RNNOISE, x);
+
+    rnnoise_process_frame(st, x, x);
+  }
+  
+  fclose(f1);
+}
+
+/*
+ * The `begin_index_element` is inclusive and the `end_index_element` is exclusive.
+ * `recording` is not modified.
+ */
+void PublishAudioWithoutNoise(int16_t recording[RECORDING_BUFFER_SIZE],
+  const long begin_index_element, const long end_index_element) {
+  vector<uint8_t> output_vector;
+
+  if (begin_index_element < end_index_element ) {
+    // The recording doesn't go circular.
+    output_vector.assign(
+      (uint8_t*)(recording + begin_index_element),
+      (uint8_t*)(recording + end_index_element));
+  } else {
+    // The recoding goes circular and we have to save it in two steps.
+    const long length_int8_elements = 
+      (RECORDING_BUFFER_SIZE - begin_index_element + end_index_element) 
+      * sizeof(int16_t);
+    output_vector.reserve(length_int8_elements);
+
+    output_vector.insert(output_vector.end(),
+      (uint8_t*)(recording + begin_index_element),
+      (uint8_t*)(recording + RECORDING_BUFFER_SIZE));
+    output_vector.insert(output_vector.end(), 
+      (uint8_t*)recording,
+      (uint8_t*)(recording + end_index_element));
+  }
+
+  audio_common_msgs::AudioData msg_audio;
+  msg_audio.data = output_vector;
```

Comment:
No da problemas de formato con ROS ? Que podemos hacer para atacar la nota en tu commit de la compatibilidad?

Reply:
Do you have information regarding the setup process of RNNoise ? I believe including it in the repo is necessary.

Reply:
> No da problemas de formato con ROS ? Que podemos hacer para atacar la nota en tu commit de la compatibilidad?

I actually don't know. I am going to do the complete integration in a following pull request.



> Do you have information regarding the setup process of RNNoise ? I believe including it in the repo is necessary.

It is actually "just" a library. Anyway, I am going to include about it in my next pull request.